open_lm/model.py

```diff
@@ -76,19 +76,61 @@ class Params:
     ffn_type: str = "swiglu"
 
 
+def get_rectangular_mask(shape, q_seq_len, k_seq_len, device, dtype):
+    next_multiple_8 = (k_seq_len + 7) // 8 * 8  #
+    mask = torch.ones((q_seq_len, next_multiple_8), device=device, dtype=bool)
+    mask[:, -q_seq_len:] = torch.tril(mask[:, -q_seq_len:], diagonal=0)
+    return torch.zeros((*shape, q_seq_len, next_multiple_8), device=device, dtype=dtype).masked_fill(
+        ~mask, float("-inf")
+    )[:, :, :, :k_seq_len]
+
+
 def xformers_attn(queries, keys, values, is_causal):
     # xformers assumes q, k, v are [batch, seq_len, heads, embed_dim]
     mask = None
-    if is_causal:
+
+    # If queries have shape [batch, 1, heads, dim] it means there is only one query in the sequence.
+    # In this case, there is no notion of causal masking, so we can just set the mask to None.
+    # This is actually needed to get the desired behavior with seq_len=1.
+    if is_causal and queries.shape[1] == keys.shape[1]:
         mask = xops.LowerTriangularMask()
+    elif is_causal and queries.shape[1] > 1:
+        # Build causal mask that assumes queries are in the end of the sequence.
```

Comment:
@achalddave You asked to comment the assumption to consider that queries are at the end of the sequence. Did you mean somewhere else than this or more details, or this is enough?